Added .NET jobs example (#4499)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/jobs/howto-schedule-and-handle-triggered-jobs.md

In your code, set up and schedule jobs within your application.

{{< tabs ".NET" "Go" >}}

{{% codetab %}}

<!-- .NET -->

> The following .NET SDK code sample schedules the job named `prod-db-backup`. Our job data will contain information

Reviewer: ```suggestion
The following .NET SDK code sample schedules the job named `prod-db-backup`. The job data contains information
```

> about the database we're seeking to make regular backups of. Over the course of this example, we'll do the following:

Reviewer: ```suggestion
about the database that you'll be seeking to backup regularly. Over the course of this example, you'll:
```

- Define types used in the rest of the example
- Register an endpoint during application startup that handles all job trigger invocations on the service
- Register the job with Dapr

> In the following example, we'll create some records that we'll serialize and register alongside the job so the information 

Reviewer: ```suggestion
In the following example, you'll create records that you'll serialize and register alongside the job so the information 
```

is available when the job is triggered in the future:
- The name of the backup task (`db-backup`)
- The backup task's `Metadata`, including:
  - The database name (`DBName`)
  - The database location (`BackupLocation`)

> Create an ASP.NET Core project and add the latest version of `Dapr.Jobs` from NuGet. While it's not strictly necessary

Reviewer: ```suggestion
Before you start, create an ASP.NET Core project and add the latest version of `Dapr.Jobs` from NuGet. While it's not necessary
```

I'd put from "While it's not..." in this line to line 46 into a note. The format for notes within `codetabs` shortcode is as following (using my suggested edits):

```
> **Note:** While it's not necessary for your project to use the `Microsoft.NET.Sdk.Web` SDK to create jobs, as of the publish time of this documentation, only the service that schedules a job receives trigger invocations for it. Because those invocations expect a registered endpoint that can handle the job trigger and requires the `Microsoft.NET.Sdk.Web` SDK, it's recommended that you use an ASP.NET Core project for this purpose.
```

But it's just a nit :)

Author: I recalled something about the notes in codetabs from the other PR and meant to ask you about it. Good catch!

> for your project to use the `Microsoft.NET.Sdk.Web` SDK to create jobs, as of the time this documentation is authored,

Reviewer: ```suggestion
for your project to use the `Microsoft.NET.Sdk.Web` SDK to create jobs, as of the publish time of this documentation,
```

> only the service that schedules a job will receive trigger invocations for it. As those invocations expect an endpoint

Reviewer: ```suggestion
only the service that schedules a job receives trigger invocations for it. Because those invocations expect a registered endpoint
```

Author: One qualm about this one - the endpoint isn't registered with Dapr, so much as it's expected to be there because of the name of the job that's registered.

Reviewer: @WhitWaldo ah ok, thanks for clarifying - just revert that change!

> registered that can handle the job trigger, and that requires the `Microsoft.NET.Sdk.Web` SDK, it's recommended that you

Reviewer: ```suggestion
that can handle the job trigger and requires the `Microsoft.NET.Sdk.Web` SDK, it's recommended that you
```

use an ASP.NET Core project for this purpose.

> We'll start by defining some types to 

Reviewer: ```suggestion
Start by defining types to 
```

> persist our backup job data and apply our own JSON property name attributes to the properties so they're consistent 

Reviewer: ```suggestion
persist the backup job data and apply your own JSON property name attributes to the properties so they're consistent 
```

with other language examples.

```cs
//Define the types that we'll represent the job data with
internal sealed record BackupJobData([property: JsonPropertyName("task")] string Task, [property: JsonPropertyName("metadata")] BackupMetadata Metadata);
internal sealed record BackupMetadata([property: JsonPropertyName("DBName")]string DatabaseName, [property: JsonPropertyName("BackupLocation")] string BackupLocation);
```

> Next, we'll set up a handler as part of our application setup that will be called anytime a job is triggered on our

Reviewer: ```suggestion
Next, set up a handler as part of your application setup that will be called anytime a job is triggered on your
```

application. It's the responsibility of this handler to identify how jobs should be processed based on the job name provided.

> This works by registering a handler with ASP.NET Core at `/job/<job-name>` where `<job-name>` is parameterized and simply

Reviewer: ```suggestion
This works by registering a handler with ASP.NET Core at `/job/<job-name>`, where `<job-name>` is parameterized and 
```

Populate your `Program.cs` file with the following: 

```cs
using System.Text;
using System.Text.Json;
using Dapr.Jobs;
using Dapr.Jobs.Extensions;
using Dapr.Jobs.Models;
using Dapr.Jobs.Models.Responses;

var builder = WebApplication.CreateBuilder(args);
builder.Services.AddDaprJobsClient();
var app = builder.Build();

//Registers an endpoint to receive and process triggered jobs
var cancellationTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(5));
app.MapDaprScheduledJobHandler((string jobName, DaprJobDetails jobDetails, ILogger logger, CancellationToken cancellationToken) => {
  logger?.LogInformation("Received trigger invocation for job '{jobName}'", jobName);
  switch (jobName)
  {
    case "prod-db-backup":
      // Deserialize the job payload metadata
      var jobData = JsonSerializer.Deserialize<BackupJobData>(jobDetails.Payload);
      
      // Process the backup operation - we assume this is implemented elsewhere in your code
      await BackupDatabaseAsync(jobData, cancellationToken);
      break;
  }
}, cancellationTokenSource.Token);

await app.RunAsync();
```

> Finally, the job itself needs to be registered with Dapr so it can be triggered at a later point in time. This could happen

Reviewer: ```suggestion
Finally, the job itself needs to be registered with Dapr so it can be triggered at a later point in time. You can do this
```

by injecting a `DaprJobsClient` into a class and executing as part of an inbound operation to your application, but for

> our purposes, we'll just put it at the bottom of our `Program.cs` file we started above. Because we'll be using the

Reviewer: ```suggestion
this example's purposes, it'll go at the bottom of the `Program.cs` file you started above. Because you'll be using the
```

> `DaprJobsClient` we registered with dependency injection, we need to start by creating a scope so we can access it.

Reviewer: ```suggestion
`DaprJobsClient` you registered with dependency injection, start by creating a scope so you can access it.
```
